### PR TITLE
firefly-73: Added improved cube, wavelength and spectral info support

### PR DESCRIPTION
**Pull request by robyww**
Firefly-73: Added improved cube, wavelength and spectral info support
   - Added zero header passed and managed by client
   - converted wavelength projection type parsing to client
   - client now maintains both projections and wavelength data
   - Updated readout to show Wavelength info
   - Improved cube paging controls to page though HDU's and cubes
   - Show wave length in cube paging controls

_To test:_

https://irsawebdev9.ipac.caltech.edu/firefly-73-cube-updates/firefly/

Read in cube:

- Notice cube controls.
- Read in multi-HDU cubes, notice  cube controls
- read in cube with wavelength information should show in mouse readout and navigation controls

@lznakano  - can you pay attentions to:
 - `Wavelength.js` - This file contains the algorithms
 - `WavelengthInfo.js` - This file analyzes the headers and prepares the parameters.

### CODE REVIEW

**Comment by vandesai1** [comment]
I tried to display the following HAWC+ image:

bacchus1:///stage/irsa-sofia-data003/20190403_HAWC_PLUS_LEVEL4/HAWC_PLUS_LEVEL4/p4353/data/g6/F0359_HA_POL_0401193_HAWEHWPE_PMP_063-101.fits

Comments related to changes made for ticket:

1. Unrecognized Extensions: At the beginning, I see "HDU: STOKES I <>
1/15" If I read in the FITS file with python/astropy, I see there are
18 extensions, not 15. It looks like the last three are not being
recognized? They are binary tables. Is that why they are not detected?

2. Confusing Mouseover: If I hover over this text, the mouseover says
"HDU: 0, Image Count: 1/16". This information seems to contradict the
label, but maybe that's because I'm misunderstanding what is meant by
"Image Count". Second, I don't think the mouseover is useful if it
basically has the same info as what is printed in the label. In this
case, perhaps we could say: "Extension in FITS File: 1/18".

3. When I try to enter my own number in the HDU box, the delete button
seems to advance the extension by 1 instead of deleting the existing
number. If I instead highlight the number and then replace it with
another, the current entry changes to something unexpected.

Comments not directly related to changes made for the ticket:

1. Would be nice if I could copy the image label.

2. There seems to be plenty of room for the image label, but I can't see the whole thing, even if I expand the window. I can't find any way of seeing the entire image label. This can be a problem for images that have the same N number of characters.

**Comment by ejoliet** [comment]
For spectral images, i think the mouse readout gives clear information needed in the mouse readout and it's perfect!

I should pointed out that this feature is really unique among FITS visualization tool and only DS9 offer something similar but in a non user-friendly way in my opinion so kudos to this!

I've tested successfully the spectral image attached in IRSA-2590 and in Firefly-73: `spectral-image__F0316_FI_IFS_04015210_BLU_WGR_700012-700013.fits`

Now, i wanted to test the images cube but without success. The file attached in Firefly-73 supposed to be a cube of images and the 3rd axis is a number that only depends on the plane index and not on the image pixel coordinates. i didn't see the number per plane in the image toolbar, i saw the mouse readout changing but this is suppose to change if the wavelength is depending on the pixel image (in case of spectral image only, right?)

I used the file in Firefly-73 `_data_SOFIA_FIFI-LS_L4_2016-07-05_FI_F316_p3993_g10_F0316_FI_IFS_04015210_BLU_WXY_600284-700065.fits`
And was expecting something like `HDU: FLUX <> 1/3 | Plane: 50 micron <> 40/57 ` in the image toolbar but didn't see any number/unit.

There are 2 kinds of multi-extension FITS files we would like to support. 
- One contain an array of images all related by the array index or 3rd axis (wavelength for example) - and the 'number'/value of the 3rd axis corresponding to the plane/image displayed would appear next to the arrows '<>' following the design proposed in ticket Firefly-20. (was already implemented for HIPS cubes: see IRSA-1629 for algorithm/formula using header values)
Example of cube in IRSA-2369, IRSA-2591

- Another kind of files are spectral images, which have extra/alternate dimension(s) that is coded in the FITS header in a particular way described by Calabretta, et.al paper (see ref. in IRSA-1498) and used to implement the algorithm to readout the alternative axis from pixel coordinates (x,y).
Example of spectral images in SOFIA: see IRSA-2590, 
Another example can be found in the sample data from IRSA-2591.

Also, I've tried without success the Spherex spectral images produced during the Spherex demo a while ago but those didn't show the wavelength in mouse readout. 
Spherex FITS file has a proper coding of the header following Calabretta et.al. which was used to implement the algorithm. Those files have a header we use to decode and drove the implementation of IRSA-1498. @lznakano could help for the debugging or code fixing when parsing those FITS spectral images in general.

I can add those files to the ticket if you need to test the parsing of those spectral images. Would be great to support those files so we don't have to come back to it to support Spherex. @vandesai1 should also give product files from Spitzer in order to  support them if we could.

I think this PR/feature is addressing the spectral images UI and added the info in the mouse readout which is very good and key most wanted feature for visualize cube of spectral images.

For spherex demo (see `IRSA-1356-adding-readout-spherex-hack` branch), the wavelength unit was actually a link to a popup dialog where the use could choose and convert to other unit `Angstrom`, `nm`, etc. (i can raise another ticket for future improvement).

**Comment by ejoliet** [comment]
Looking again at the header of the file `_data_SOFIA_FIFI-LS_L4_2016-07-05_FI_F316_p3993_g10_F0316_FI_IFS_04015210_BLU_WXY_600284-700065.fits`, the HDU2:

```
# HDU 2 in _data_SOFIA_FIFI-LS_L4_2016-07-05_FI_F316_p3993_g10_F0316_FI_IFS_04015210_BLU_WXY_600284-700065.fits:
XTENSION= 'IMAGE   '           /extension type                                  
BITPIX  =                  -64 /bits per data value                             
NAXIS   =                    3 /number of axes                                  
NAXIS1  =                   33 /                                                
NAXIS2  =                   35 /                                                
NAXIS3  =                   49 /                                                
PCOUNT  =                    0 /                                                
GCOUNT  =                    1 /                                                
EXTNAME = 'ERROR   '           /                                                
DATE-OBS= '2016-07-05T14:17:46' / Observation date                              
BUNIT   = 'Jy      '           / Data units                                     
EQUINOX =              2000.00 / Coordinate equinox                             
CTYPE1  = 'RA---TAN'           / Axis 1 type and projection                     
CTYPE2  = 'DEC--TAN'           / Axis 2 type and projection                     
CTYPE3  = 'WAVE    '           / Axis 3 type and projection                     
CUNIT1  = 'deg     '           / Axis 1 units                                   
CUNIT2  = 'deg     '           / Axis 2 units                                   
CUNIT3  = 'um      '           / Axis 3 units                                   
CRPIX1  =              1.00000 / Reference pixel (x)                            
CRPIX2  =              1.00000 / Reference pixel (y)                            
CRPIX3  =              1.00000 / Reference pixel (z)                            
CRVAL1  =        253.223795141 / RA (deg) at CRPIX1,2                           
CRVAL2  =       -43.9107844345 / Dec (deg) at CRPIX1,2                          
CRVAL3  =        63.0202669160 / Wavelength (um) at CRPIX3                      
CDELT1  =   -0.000277777777778 / RA pixel scale (deg/pix)                       
CDELT2  =    0.000277777777778 / Dec pixel scale (deg/pix)                      
CDELT3  =     0.00718111003853 / Wavelength pixel scale (um/pix)                
CROTA2  =              0.00000 / Rotation angle (deg) 
```


It seems to indicate that there are 49 images, and the wavelength per plane is a `value = crval3 + ( plane-number - crpix3 ) * cdelt3`.

In comparison with spectral images where `NAXIS3 = 1` or is not even present, because the coding of the information is based on the transformation from pixel coordinate (x,y) to z using CRPIX1,2, CRVAL1,2 and CDELT1,2. Example from SOFIA :

```
# HDU 1 in spectral-image__F0316_FI_IFS_04015210_BLU_WGR_700012-700013.fits:
XTENSION= 'IMAGE   '           /extension type                                  
BITPIX  =                  -64 /bits per data value                             
NAXIS   =                    2 /number of axes                                  
NAXIS1  =                   49 /                                                
NAXIS2  =                   25 /                                                
PCOUNT  =                    0 /                                                
GCOUNT  =                    1 /                                                
EXTNAME = 'FLUX    '           / `

Where the `lambda` is defined in the main header:
`CRVAL3  =        63.1937683390 / W at ref. pixel                                
CRPIX3  =              24.5000 / Reference pixel number for LAMBDA              
CTYPE3  = 'LAMBDA  '           /                                                
CDELT3  =           0.00718111 / Lambda grid spacing  
```

And spherex example (i've attached the file in the ticket Firefly-73):
```
# HDU 0 in testLinear.fits:
SIMPLE  =                    T / conforms to FITS standard                      
BITPIX  =                  -64 / array data type                                
NAXIS   =                    2 / number of array dimensions                     
NAXIS1  =                 2048                                                  
NAXIS2  =                 2048                                                  
WCSAXES =                    3 / Number of coordinate axes                      
CRPIX1  =               1024.0 / Pixel coordinate of reference point            
CRPIX2  =               1024.0 / Pixel coordinate of reference point            
CRPIX3  =                  0.0 / Pixel coordinate of reference point            
PC1_1   =            -0.001389 / Coordinate transformation matrix element       
PC2_2   =             0.001389 / Coordinate transformation matrix element       
PC3_1   =             2.93E-10 / Coordinate transformation matrix element       
PC3_2   =                  1.5 / Coordinate transformation matrix element       
PC3_3   =                  3.5 / Coordinate transformation matrix element       
CDELT1  =                  1.0 / [deg] Coordinate increment at reference point  
CDELT2  =                  1.0 / [deg] Coordinate increment at reference point  
CDELT3  =                  1.0 / [m] Coordinate increment at reference point    
CUNIT1  = 'deg'                / Units of coordinate increment and value        
CUNIT2  = 'deg'                / Units of coordinate increment and value        
CUNIT3  = 'm'                  / Units of coordinate increment and value        
CTYPE1  = 'RA---TAN'           / Right ascension, gnomonic projection           
CTYPE2  = 'DEC--TAN'           / Declination, gnomonic projection               
CTYPE3  = 'WAVE'               / Vacuum wavelength (linear)                     
CRVAL1  =                150.0 / [deg] Coordinate value at reference point      
CRVAL2  =                  2.0 / [deg] Coordinate value at reference point      
CRVAL3  =              7.5E-07 / [m] Coordinate value at reference point        
LONPOLE =                180.0 / [deg] Native longitude of celestial pole       
LATPOLE =                  2.0 / [deg] Native latitude of celestial pole        
WCSAXISB=                    2                                                  
CRPIX1B =                  0.0                                                  
CDELT1B =             2.93E-10                                                  
CUNIT1B = 'm       '                                                            
CTYPE1B = 'WAVE    '                                                            
CRVAL1B =              7.5E-07                                                  
CRPIX2B =                 1024                                                  
CDELT2B =             0.001389                                                  
CUNIT2B = 'deg     '                                                            
CTYPE2B = 'DEC--TAN'                                                            
CRVAL2B =                  2.5
```

Example from Spitzer when the extra axis is defined in a table lookup:
```
# HDU 0 in ngc0337_DR5_LH_cube.fits:
SIMPLE  =                    T /Written by IDL:  Fri Mar 30 11:34:07 2007       
BITPIX  =                  -32 /Real*4 (floating point)                         
NAXIS   =                    3 /                                                
NAXIS1  =                   10 /                                                
NAXIS2  =                    7 /                                                
NAXIS3  =                  880 /                                                
EXTEND  =                    T /File contains extensions                        
DATE    = '2007-03-30'         /                                                
APERNAME= 'LH      '           / IRS module and order                           
BUNIT   = 'MJy/sr  '           / Units of flux data                             
CAL_SET = 'irs_2007_02_12-pb-pfc-trim-omeg-lhbiasfork.cal' / Calibration set    
SOFTWARE= 'CUBISM  '           / Product assembly software                      
SOFT_VER= 'v1.50pre3 (March-2007)' / Software version used                      
CTYPE1  = 'RA---TAN'           / Coordinate Type                                
CTYPE2  = 'DEC--TAN'           / Coordinate Type                                
EQUINOX =              2000.00 / Equinox of Ref. Coord.                         
PC1_1   =       0.356597772325 / Degrees / Pixel                                
PC2_1   =      -0.934258009745 / Degrees / Pixel                                
PC1_2   =       0.934258009745 / Degrees / Pixel                                
PC2_2   =       0.356597772325 / Degrees / Pixel                                
CDELT1  =    -0.00123888894450 / Degrees / Pixel                                
CDELT2  =     0.00123888894450 / Degrees / Pixel                                
CRPIX1  =              5.50000 / Reference Pixel in X                           
CRPIX2  =              4.00000 / Reference Pixel in Y                           
CRVAL1  =        14.9576714702 / R.A. (degrees) of reference pixel              
CRVAL2  =       -7.57797923055 / Declination of reference pixel                 
LONPOLE =        180.000000000 / Native longitude of Celestial pole             
LATPOLE =        0.00000000000 / Celestial latitude of native pole              
PV2_1   =        0.00000000000 /Projection parameter 1                          
PC3_3   =                    1 / Index Separation in wavelength table           
CRPIX3  =                    1 / Reference pixel                                
CRVAL3  =                    1 / Table index for reference pixel                
CTYPE3  = 'WAVE-TAB'           / Wavelength                                     
CUNIT3  = 'um      '           / Wavelength units                               
PS3_0   = 'WCS-TAB '           / Coordinate table extension name                
PS3_1   = 'WAVELENGTH'         / Coordinate table column name`
....

# HDU 1 in ngc0337_DR5_LH_cube.fits:
XTENSION= 'BINTABLE'           /Written by IDL:  Fri Mar 30 11:34:07 2007       
BITPIX  =                    8 /                                                
NAXIS   =                    2 /Binary table                                    
NAXIS1  =                 3520 /Number of bytes per row                         
NAXIS2  =                    1 /Number of rows                                  
PCOUNT  =                    0 /Random parameter count                          
GCOUNT  =                    1 /Group count                                     
TFIELDS =                    1 /Number of columns                               
EXTNAME = 'WCS-TAB '           /Wavelength look-up table for cube dimension 3   
TFORM1  = '880E    '           /Real*4 (floating point)                         
TTYPE1  = 'WAVELENGTH'         /Column label                                    
TDIM1   = '(1,880) '           /Array dimensions for column 1                   
TUNIT1  = 'um      '           /Units of column 1 
```
As you can see there are many formats and this is only a tiny sample.
Initially we would like to support SOFIA spectral images and then Spherex/Spitzer/??.

I really think we need the support of @vandesai1 and @gpdf here to give the formats and the data samples for testing. 

Another one i used in the past (don't know where i got it), it's a spectral cube with 2 extra axis:

```
# HDU 0 in ARP_220_I_6cm_n1988.fits.gz:
SIMPLE  =                    T /                                                
BITPIX  =                  -32 /                                                
NAXIS   =                    4 /                                                
NAXIS1  =                  256 /                                                
NAXIS2  =                  256 /                                                
NAXIS3  =                    1 /                                                
NAXIS4  =                    1 /                                                
EXTEND  =                    T /Tables following main image                                           
OBJECT  = 'IC4553  '                                                          
CTYPE1  = 'RA---SIN'                                                            
CRVAL1  =    2.33195414543E+02 /                                                
CDELT1  =     -8.333333426E-06 /                                                
CRPIX1  =      1.280000000E+02 /                                                
CROTA1  =      0.000000000E+00 /                                                
CTYPE2  = 'DEC--SIN'                                                            
CRVAL2  =    2.36688885838E+01 /                                                
CDELT2  =      8.333333426E-06 /                                                
CRPIX2  =      1.290000000E+02 /                                                
CROTA2  =      0.000000000E+00 /                                                
CTYPE3  = 'FREQ    '                                                            
CRVAL3  =    4.88500019200E+09 /                                                
CDELT3  =      5.000000000E+07 /                                                
CRPIX3  =      1.000000000E+00 /                                                
CROTA3  =      0.000000000E+00 /                                                
CTYPE4  = 'STOKES  '                                                            
CRVAL4  =    1.00000000000E+00 /                                                
CDELT4  =      1.000000000E+00 /                                                
CRPIX4  =      1.000000000E+00 /                                                
CROTA4  =      0.000000000E+00 / ```

BTW, Firefly upload doesn't accept any extension, only `.fits`. So in order to upload this last sample file. i had to rename it from `.fits.gz` to `.fits`. I think in the past we were not checking the extension, did we?

**Comment by robyww** [comment]
So now that all the infrastructure is there for parsing.  We just need to go through the example data and add what is missing.  I will take a look and get all the ones that are not working implemented. The hard part is over.

**Comment by ejoliet** [comment]
Sounds great!

**Comment by ejoliet** [comment]
BTW, in order to inspect (or play with) a FITS file, there is an OLD tool 'fv' from HEASARC that actually 'digest' anything.

While i was investigating those FITS files, i thought we could similarly as part of a future improvement, display a plot when the HDU is actually an 'image' with one axis, instead of trying to display it as an image or something else. In the upload or anywhere else we are dealing with FITS, we could give 3 options: display table, plot or image.

Same for 2d 'images' that are actually 2d spectra, we could just display the table as if it was a binary table. See https://imgur.com/a/cEOMtWz
Once you click on plot, it will read the 'image' binary pixels and covert the value and plot it as pixel value vs row index, for example `TRANSMISSION` 1x49 looks like: https://imgur.com/a/d5KRiE6

Fv tool can be install from here:https://heasarc.gsfc.nasa.gov/lheasoft/ftools/fv/fv_download.html

**Comment by robyww** [comment]
> Unrecognized Extensions: At the beginning, I see "HDU: STOKES I <>
1/15" If I read in the FITS file with python/astropy, I see there are
18 extensions, not 15. It looks like the last three are not being
recognized? They are binary tables. Is that why they are not detected?

The image viewer part of Firefly just show the image extension.  So this is correct. If we want to change how this  works there is are  a lot of issue to consider.  It should be pursued in a separate ticket. BTW- the upload dialog give a complete analysis of the file.  That is what it is for.


> Confusing Mouseover: If I hover over this text, the mouseover says
"HDU: 0, Image Count: 1/16". This information seems to contradict the
label, but maybe that's because I'm misunderstanding what is meant by
"Image Count". Second, I don't think the mouseover is useful if it
basically has the same info as what is printed in the label. In this
case, perhaps we could say: "Extension in FITS File: 1/18".

Image Count is the total number of images in the file. If there we 4 HDUs, cube:5, cube:5, cube:3, single image.  The the image count would be 14. We can re-label it in the tool tip but it is useful information.

> When I try to enter my own number in the HDU box, the delete button
seems to advance the extension by 1 instead of deleting the existing
number. If I instead highlight the number and then replace it with
another, the current entry changes to something unexpected.

I can tweek this this some.  it is made to show immediate results and not wait for you to hit enter.  I do see a bug when you try to delete the number completely. However when working correctly it you enter "12" it should show image #1 and the #12.

**Comment by lznakano** [comment]
Here is what my reviewing results:
1. Test this data (_data_SOFIA_FIFI-LS_L4_2016-07-05_FI_F316_p3993_g10_F0316_FI_IFS_04015210_BLU_WXY_600284-700065.fits) and it worked fine.
2. Test this data (spectral-image__F0316_FI_IFS_04015210_BLU_WGR_700012-700013.fits), it showed wavelength but there is no coordinate information displayed.  
3. I tested F2W, V2W and log with my simulated FITs files and they are not working.  The reason is that the WCSAXES keyword is not processed.  In each simulated FITs file, it has a 2-d image with wavelength information specified by keywords.  I am not sure if we want to support this kind format or not.  The Java version was implemented according to the reference paper (Grisen's paper) in the ticket.  

Other than those, everything else is great!!!

**Comment by vandesai1** [comment]
4. F0316_FI_IFS_04015210_RED_WXY_700028-700065.fits - This is a MEF
that contains a cube. The wavelength is not listed with the Plane information, which is where I would have expected it. Instead, it is in the readout area. Also, the units for the wavelength are listed as
just the Greek letter mu, rather than "micron" or "um". Finally, and most important, the wavelength is varying across the image, and I don't think it is supposed to. I think there is supposed to be just one wavelength per plane.

**Comment by vandesai1** [comment]
> @vandesai1 :
> 
> > Unrecognized Extensions: At the beginning, I see "HDU: STOKES I <>
> > 1/15" If I read in the FITS file with python/astropy, I see there are
> > 18 extensions, not 15. It looks like the last three are not being
> > recognized? They are binary tables. Is that why they are not detected?
> 
> The image viewer part of Firefly just show the image extension. So this is correct. If we want to change how this works there is are a lot of issue to consider. It should be pursued in a separate ticket. BTW- the upload dialog give a complete analysis of the file. That is what it is for.
> 

I was only using the Image search because we haven't ingested all of the SOFIA data into the SOFIA application. When the application shows an EXES file, the user hasn't specified that they only want to see images. So we should report all extensions. I can make that a separate ticket if need be.

> > Confusing Mouseover: If I hover over this text, the mouseover says
> > "HDU: 0, Image Count: 1/16". This information seems to contradict the
> > label, but maybe that's because I'm misunderstanding what is meant by
> > "Image Count". Second, I don't think the mouseover is useful if it
> > basically has the same info as what is printed in the label. In this
> > case, perhaps we could say: "Extension in FITS File: 1/18".
> 
> Image Count is the total number of images in the file. If there we 4 HDUs, cube:5, cube:5, cube:3, single image. The the image count would be 14. We can re-label it in the tool tip but it is useful information.
> 
Try loading the following file:

bacchus1://stage/irsa-sofia-data003/20190403_HAWC_PLUS_LEVEL4/HAWC_PLUS_LEVEL4/p4353/data/g6/F0359_HA_POL_0401193_HAWEHWPE_PMP_063-101.fits

There are 15 HDUs, and none of them are cubes. Yet the mouseover says that the Image Count is 16. I don't understand that.

> > When I try to enter my own number in the HDU box, the delete button
> > seems to advance the extension by 1 instead of deleting the existing
> > number. If I instead highlight the number and then replace it with
> > another, the current entry changes to something unexpected.
> 
> I can tweek this this some. it is made to show immediate results and not wait for you to hit enter. I do see a bug when you try to delete the number completely. However when working correctly it you enter "12" it should show image #1 and the #12.

That would be great, but I cannot enter any number, either '1' or '12'.

**Comment by robyww** [comment]
> bacchus1://stage/irsa-sofia-links/archive/FIFI-LS/L4/2016-07-05_FI_F316/p3993/g11/F0316_FI_IFS_04015210_RED_WXY_700028-700065.fits

> This is a MEF that contains a cube. Each plane in the cube represents a different wavelength. Problems I see:
> The wavelength is not listed with the Plane information, which is where I would have expected it based on the description above. Instead, it is in the readout area. The parsing algorithm the we are using have this at the pixel level

That is not what we are recognizing.  If it is different from our algorithm I need to know how. When `CTYPE3 = WAVE` we are looking for pixel level changes. `CRVAL3 + CDELT3*(x+y)` or a more complex one `PC3_` and `CD3_i`
when `CTYPE3` is not defined then  `crval3 + ( cubeIdx - crpix3 ) * cdelt3` 

> The units for the wavelength are listed as just the Green letter mu, rather than "micron" or "um".

I will fix this it should be `µm`

> The displayed wavelength varies with pixel coordinates. Instead, it should vary with plane.

This is something the we need to understand then.

Right now the plan is  for plane level wavelength to show it in both places on the control and on the mouse readout.  If it is pixel level wavelength then just show it on the mouse readout

**Comment by gpdf** [comment]
@robyww If a cube (that is, an image extension with NAXIS=3, even if NAXIS3=1) has "normal" WCS type values for axis 1 and axis 2 (e.g., CTYPE1/2 = RA---TAN/DEC---TAN) and then has a 3rd-axis type like CTYPE3 = WAVE, that means that it is NAXIS3 separate images, one per plane, with identical spatial WCSes and a single, constant wavelength per plane.  The pixel-coordinates field of the UI should NOT be used to display wavelength in this case; it should appear in the image's title bar.

Even if NAXIS3=1 (where there's only a single image) the appropriate wavelength should be displayed in the image title bar: if the data producer has gone to the trouble of labeling their data in this way, we should respect that.  However, this is a lower priority than getting this right for cubes with more than one plane.

**Comment by gpdf** [comment]
The pixel-coordinate-area wavelength readout should only be used when there is a mixed spatial/wavelength WCS.  This is what we'll have with, for instance, SPHEREx Level 1 images.

An example of this is in WCS "B" in testLinear.fits above, where you have

```
CUNIT1B = 'm       '                                                            
CTYPE1B = 'WAVE    '                                                            
...
CUNIT2B = 'deg     '                                                            
CTYPE2B = 'DEC--TAN'
```

**Comment by vandesai1** [comment]
2. Test this data (spectral-image__F0316_FI_IFS_04015210_BLU_WGR_700012-700013.fits), it showed wavelength but there is no coordinate information displayed.

There is no coordinate information displayed because these data have not been spatially calibrated, yet. 
Also, although wavelength info is shown, I am not sure it is correct. According to IRSA-2590, the wavelength is supposed to vary along the rows, not along the columns.

**Comment by robyww** [comment]
Update from feedback

  - Fixed µm
  - added enum option to fits header table, and some rewording
  - Wavelength will prefer a alt-wcs (i.e B) if on exist
  - improve the plane level wavelength calculation
  - Wavelength readout will only be in mouse readout section if a single plane or has pixel level wavelength data
  - fixed UI cube nav control
  - support TAB, passing the table, code ported from java

**Comment by robyww** [comment]
_Note on HDU's_

The Firefly client does not know about HDU’s that are not 2-d images or cubes. Therefore if there are 10 HDU's and 5 are images.  Firefly sees only the 5 images. It does know the original HDU number. To enhance the way it is going to have to be in another ticket. It is a decent amount of work. The count on the image pager is the number of 2-d or cubes not the number of HDUs.  

We could change image image pager to show the original HDU number somewhere but right now the HDU number entry needs to be consecutive.   

There is also the bigger picture of what we want to happen if the HDU pages chooses a non-image HDU.

**Comment by robyww** [comment]
After some though and support from @ejoliet I will merge this ticket.  If there are any more UI issues that we really want fixed for this release I will make a new ticket. If we encounter some show-stopper issue with wavelength readouts it is very easy to disable.

 I created FIREFLY-116 to evaluate the scientific accuracy and fix.
